In [10]:
#url = "https://docs.google.com/spreadsheets/d/1eZj7EOdGe16P5QWHSIIHhkclNAog13e7xF6fPMoXfEY/edit#gid=1897262175"
name = "main"

import pandas as pd
import urllib.parse

sheet_id = "1eZj7EOdGe16P5QWHSIIHhkclNAog13e7xF6fPMoXfEY"
sheet_name = "Buy_Insta_Accounts"
#sheet_name_encoded = urllib.parse.quote(sheet_name)
url = f"https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet={sheet_name}"

In [14]:
fn = "Anthony MainSheet [current projects] - Buy_Insta_Accounts.tsv"
Buy_Insta_Accounts_df = pd.read_csv(fn, sep="\t")

## Load data

In [2]:
import os 
import pandas as pd

files = os.listdir("data")

df_women_event = pd.read_csv(os.path.join("data", files[0]), sep="\t")
df_buy_insta_accounts_add_on = pd.read_csv(os.path.join("data", files[1]), sep="\t")
df_buy_biz = pd.read_csv(os.path.join("data", files[2]), sep="\t")
df_buy_insta_accounts = pd.read_csv(os.path.join("data", files[3]), sep="\t")
df_women_event_mined = pd.read_csv(os.path.join("data", files[4]), sep="\t")

## EDA

In [85]:
df_buy_biz.media_count.describe()

count     143.000000
mean      405.426573
std       583.785926
min         0.000000
25%        54.000000
50%       166.000000
75%       481.500000
max      3813.000000
Name: media_count, dtype: float64

In [31]:
df_women_event_mined.columns

Index(['username', 'userid', 'scraped_from', 'media_count', 'full_name',
       'following_count', 'follower_count', 'first_name', 'external_url',
       'email', 'contact_phone_number', 'cityName', 'category', 'biography',
       'address_street', 'Big 5'],
      dtype='object')

In [38]:
df_buy_insta_accounts.columns

Index(['username', 'Openness', 'Conscientiousness', 'Extroversion',
       'Agreeableness', 'Neuroticism', 'Type', 'media_count', 'full_name',
       'following_count', 'follower_count', 'first_name', 'external_url',
       'engagement_percent', 'email', 'contact_phone_number', 'cityName',
       'category', 'biography', 'address_street', 'AvgLikes', 'AvgComments'],
      dtype='object')

In [50]:
df_buy_biz[df_buy_biz['Openness'].notna()]

username Openness Conscientiousness Extroversion  \
0               goodeatzco      Low              High         High   
2               a1republic     High              High         High   
3           keychainsocial     High              High         High   
4        thegiftshoppe.nft    Vague               Low          Low   
5       officialclubsocial      Low              High         High   
..                     ...      ...               ...          ...   
138           irose.social     High              High         High   
139               nattcity     High               Low         High   
140             jondennill     High              High         High   
141            igmodels_co      Low              High        Vague   
142  skystrategicmarketing     High              High         High   

    Agreeableness Neuroticism      Type  media_count                full_name  \
0           Vague       Vague      Blog         1693       #GoodEatzCo | Andy   
2            High         Low  Business           30            A1 REPUBLIC ™   
3            High         Low  Business          214    Keychain Social, Inc.   
4           Vague       Vague  Business            9          the gift shoppe   
5           Vague       Vague  Business            6              Club Social   
..            ...         ...       ...          ...                      ...   
138           Low         Low  Business          338                    IROSE   
139           Low         Low  Personal         1129                 nathalie   
140          High         Low  Personal          124         Jonathan Dennill   
141         Vague       Vague      Blog          458    IG Models Worldwide ®   
142          High         Low  Business          427  Sky Strategic Marketing   

     following_count  ...  engagement_percent  \
0                145  ...                0.14   
2                613  ...                0.53   
3                  0  ...                 NaN   
4                  5  ...                0.57   
5                  2  ...                 NaN   
..               ...  ...                 ...   
138              422  ...                0.30   
139              686  ...                0.06   
140              733  ...                0.49   
141              158  ...                0.01   
142             2133  ...                0.18   

                               email contact_phone_number  \
0               goodeatzco@gmail.com                  NaN   
2               a1republic@yahoo.com         1.949934e+10   
3            info@keychainsocial.com                  NaN   
4        hello@fromthegiftshoppe.com                  NaN   
5           clubsocialmgmt@gmail.com                  NaN   
..                               ...                  ...   
138           sharon@irosesocial.com                  NaN   
139  nathalia@nfluenceconsulting.com                  NaN   
140            jon.dennill@gmail.com                  NaN   
141            igmodels.co@gmail.com                  NaN   
142  tanya@skystrategicmarketing.com         1.813337e+10   

                      cityName             category  \
0      Los Angeles, California              Blogger   
2    Newport Beach, California  Social Media Agency   
3      Los Angeles, California                  NaN   
4                          NaN                  NaN   
5      Los Angeles, California  Social Media Agency   
..                         ...                  ...   
138    Los Angeles, California      Product/service   
139                        NaN                Actor   
140                        NaN                  NaN   
141    Los Angeles, California  Social Media Agency   
142             Tampa, Florida                  NaN   

                                             biography  \
0    🍴Professional Food Consumer🤘🏻😛 Tiktok 🎥: @good...   
2    The #1 organic IG growth service for influence...   
3    Digital Social Media Agency • Unlock Your 

In [52]:
len(df_buy_insta_accounts_add_on)

2734

In [35]:
df_buy_insta_accounts_add_on

username Openness Conscientiousness Extroversion  \
0        patrickbreenmusic      Low              High          Low   
1     love.lindsey.wheeler     High              High         High   
2             authorkwells     High               Low         High   
3        aleksatodorovicev     High              High         High   
4         grigoruadisveley     High              High         High   
...                    ...      ...               ...          ...   
2729        danielarendonn      NaN               NaN          NaN   
2730         the_fifthguys      NaN               NaN          NaN   
2731              mskialee      NaN               NaN          NaN   
2732        beckie_farrant      NaN               NaN          NaN   
2733          mikelonghelt      NaN               NaN          NaN   

     Agreeableness Neuroticism      Type  media_count  \
0             High         Low  Personal          504   
1             High         Low  Personal         2381   
2             High         Low  Personal          129   
3              Low         Low  Personal          897   
4              Low         Low  Personal        10742   
...            ...         ...       ...          ...   
2729           NaN         NaN       NaN          408   
2730           NaN         NaN       NaN          167   
2731           NaN         NaN       NaN          236   
2732           NaN         NaN       NaN         2109   
2733           NaN         NaN       NaN          987   

                          full_name  following_count  ...  \
0                     Patrick Breen              893  ...   
1                   LINDSEY WHEELER             1332  ...   
2     Author Kamisha Wells RDKS CEO              314  ...   
3                 Todorović Aleksa              876  ...   
4       Anzhelika Grigoryadis Veley              980  ...   
...                             ...              ...  ...   
2729                 Daniela Rendon               36  ...   
2730                  The FifthGuys              420  ...   
2731          Ms. Lee/ Royalty حقية              102  ...   
2732                 Beckie Farrant             1339  ...   
2733                   mikelonghelt              539  ...   

                                           external_url engagement_percent  \
0                                                   NaN              2.793   
1                          https://www.lovelindsey.com/              1.063   
2              https://www.amazon.com/shop/authorkwells              1.804   
3                   https://linktr.ee/aleksatodorovicev              0.703   
4                   https://paysend.com/referral/01g6gi              0.470   
...                                                 ...                ...   
2729                   https://linktr.ee/danielarendonn              0.674   
2730                      https://discord.gg/CpdVpPaZXu              0.777   
2731                         https://linktr.ee/MSKIALEE              0.045   
2732  https://www.greatstuff.dupont.com/great-pantry...              0.498   
2733                             https://ffm.to/mmlqzdw              0.274   

                                 email  contact_phone_number  \
0        patrick@patrickbreenmusic.com                   NaN   
1            lindseyslularoe@gmail.com          +17576206307   
2               kamishawells@yahoo.com          +15627350110   
3            todorovicaskela@gmail.com                   NaN   
4     anzhelika-grigorjadis@rambler.ru                   NaN   
...                                ...                   ...   
2729         daniela@danielarendon.com          +17868997656   
2730    thefifthguysofficial@gmail.com                   NaN   
2731              klee@keybridgeus.com          +17706424307   
2732     infarrantlycreative@gmail.com                   NaN   
2733               mlonghelt@gmail.com                   NaN   

                     cityName         category  \
0  

## Meaning of the columns
https://developers.facebook.com/docs/instagram-api/reference/ig-user/
This URL has field defintion.
* biography - Profile bio text.
* followers_count - Total number of Instagram users following the user.
* following_count - Total number of Instagram users the user follows.
* media_count - Total number of IG Media published on the user.


In [6]:
set(df_women_event_mined.username)

0        hannahmariethemuse
1            refinedlash.co
2         studio_irene_rose
3              alena_smily9
4        sincerelyluna.xoxo
                ...        
11276                lianev
11277                   kuz
11278           tanamongeau
11279             lilyachty
11280             jenselter
Name: username, Length: 11281, dtype: object

In [12]:
df_women_event

username  media_count                      full_name  \
0             ariannyceleste         6039       Arianny Celeste UFC Mama   
1                 sarahstage         1501                    Sarah Stage   
2     beauty_nurse_elizabeth         1060           Elizabeth Sommers RN   
3                  carlyrbel          493                      CARLY BEL   
4              houdaschimmel          507                 Houda Schimmel   
...                      ...          ...                            ...   
1467            houghtonmade            1  Memorable and Detailed Events   
1468               violacomm            3           Viola Communications   
1469  brainstorm.media.group            0         Brainstorm Media Group   
1470     littlestudio_beauty           33           Little Studio_Beauty   
1471           adtop1percent            0   Ann Dee | LA ENTREPRENEUR LA   

      following_count  follower_count first_name  \
0                 997         3288148    Celeste   
1                1472         2086691      Sarah   
2                 383         1504557  Elizabeth   
3                 191         1035410      Carly   
4                3041          693036      Houda   
...               ...             ...        ...   
1467              166              21     Lauren   
1468                3              20      Viola   
1469               32              20  Katherine   
1470              146              12   Laetitia   
1471               42               7        Ann   

                          external_url  engagement_percent  \
0     http://www.ariannyexclusive.com/                0.06   
1         https://linktr.ee/Sarahstage                0.01   
2     http://www.elizabethsommers.com/                0.23   
3                 http://carlybel.com/                0.14   
4                                  NaN                 NaN   
...                                ...                 ...   
1467                               NaN               10.95   
1468   http://violacommunications.com/                6.83   
1469  http://brainstormmediagroup.com/                0.00   
1470       http://www.littlestudio.us/                4.03   
1471                               NaN                0.00   

                                   email contact_phone_number  \
0             arianny@ariannyceleste.com                  NaN   
1               booksarahstage@gmail.com                  NaN   
2                    sommersrn@gmail.com                  NaN   
3                carlybelsnaps@gmail.com                  NaN   
4               infos@shimmypetstore.com                  NaN   
...                                  ...                  ...   
1467             laurenjmadrid@gmail.com         +13234008726   
1468        info@violacommunications.com                  NaN   
1469  katherine@brainstormmediagroup.com                  NaN   
1470            laetitia@littlestudio.fr                  NaN   
1471             adtop1percent@gmail.com                  NaN   

                     cityName       category  \
0     Los Angeles, California  Public figure   
1                         NaN            NaN   
2     Los Angeles, California  Health/beauty   
3                         NaN   Entrepreneur   
4     Los Angeles, California   Entrepreneur   
...                       ...            ...   
1467  Los Angeles, California            NaN   
1468           Tampa, Florida            NaN   
1469  Los Angeles, California            NaN   
1470                      NaN            NaN   
1471                      NaN   Entrepreneur   

                                              biography address_street  \
0     Raidens Mom 👶🏽🐶🐶🧿 Founder @girlfriendbox @them...            NaN   
1     Boy Mom 👶🏻👦🏻 Wife 💃🏻 & Online Fitness Coach 💪🏼...            NaN   
2     Aesthetic RN-BSN •Injection Artist 🌹 •Natural ...            NaN   
3     One of a kind ✨ • @carlybelx • shop @carlyclub...            NaN   
4        

In [9]:
len(set(df_women_event_mined.username).intersection(set(df_women_event.username)))

3

In [10]:
len(set(df_buy_insta_accounts.username).intersection(set(df_buy_insta_accounts_add_on.username)))

0

## Call Instagram Business Discovery API

`df_buy_insta_accounts` has 94 rows with big 5 score, total 5535 rows

`df_buy_biz` total 143 rows, 129 rows have big 5 scores

`df_buy_insta_accounts_add_on` 2734 rows, 95 have big 5 scores

=======

`df_women_event_mined` Big 5 column is empty, total 11281 rows


`df_women_event` 1472 rows, no big five scores

In [6]:
import reqxests
import json
import pandas as pd
def get_media_metadata(ig_user_name, token, pages=1):
    params = {
        "ig_username": ig_user_name,
        "access_token": token,
        "endpoint_base":"https://graph.facebook.com/v17.0/",
        "instagram_account_id":"17841400224190654"
    }
    biz_discovery_first_page = '){username,ig_id,id,profile_picture_url,biography,follows_count,followers_count,media_count, media{comments_count, like_count, caption, timestamp, media_type}}' 

    endpointParams = dict() # parameter to send to the endpoint
    
    endpointParams['fields'] = 'business_discovery.username(' + params['ig_username'] + biz_discovery_first_page
    
    # string of fields to get back with the request for the account
    endpointParams['access_token'] = params['access_token'] # access token

    url = params['endpoint_base'] + params['instagram_account_id'] # endpoint url
    
    data = requests.get(url, endpointParams)
    if data.ok:
        dict_ = json.loads(data.content)
    else:
        return data.reason
    try:
        df = pd.DataFrame(dict_['business_discovery']['media']['data'])
    except:
        print(f"{ig_user_name} has bad data")
        return dict_
    
    pages -=1
    while pages>0:
        #d1 = json.loads(data.content)
        try:
            cursor = dict_['business_discovery']['media']['paging']['cursors']['after']
        except:
            return "no cursor"
        biz_discovery_following_page = f"){{username,website,name,ig_id,id,profile_picture_url,biography,follows_count,followers_count,media_count, media.after({cursor}){{comments_count, like_count, caption, timestamp, media_type}}}}"
        endpointParams['fields'] = 'business_discovery.username(' + params['ig_username'] + biz_discovery_following_page
        url = params['endpoint_base'] + params['instagram_account_id'] # endpoint url
        data = requests.get(url, endpointParams)
        dict_ = json.loads(data.content)
        df = pd.concat([df, pd.DataFrame(dict_['business_discovery']['media']['data'])])
                       
        pages -=1
        
    
    return df


In [30]:
from tqdm import tqdm
import os
from pathlib import Path

# go to: https://developers.facebook.com/tools/explorer/
# to get token
# this token is good till August 17th
token ="EAACccrMmWa0BAGziHtFogA885lMeFU3LHBHXDqZBtAMiTdmCUw7WLprdQUMMWKFZCjqhnlcDhZC2nIYFPzohpRsLBaaTopC4n2FAmqwbrTbJZBTixRBEhBlrrhl80tF3apZC2wEEWaj6bzVKKMepM4990lzugk5kWZB8egWVc8ZAgZDZD"

def get_biz_discovery_e2e(usernames, folder):
    
    try:
        os.mkdir(f"feature/{folder}")
    except:
        pass

    scraped_users = []
    for entry in Path(f"feature/{folder}").iterdir():
        # check if it a file
        if entry.is_file():
            scraped_users.append(entry.stem)

    bad_users = []
    for ig_username in tqdm(usernames.tolist()):
        if ig_username in scraped_users:
            continue

        res = get_media_metadata(ig_username, token, 1)
        if type(res) is str:
            if res == "Forbidden":
                print("possibly reach rate limit")
                break
            bad_users.append({ig_username: res})
        else:
            res.to_parquet(f"feature/{folder}/{ig_username}.parquet")
    return bad_users, scraped_users

In [31]:
usernames = df_buy_insta_accounts[df_buy_insta_accounts['Openness'].isna()]['username'][300:500]
folder = "buy_insta_accounts"
bad_users, scraped_users = get_biz_discovery_e2e(usernames, folder)

 98%|█████████▊| 195/200 [00:04<00:00, 45.75it/s] 

possibly reach rate limit


In [27]:
len(bad_users)

17

In [23]:
bad_users

[{'daniellevy01': 'Bad Request'},
 {'alannajadeyoga': 'Bad Request'},
 {'punkerstv': 'Bad Request'},
 {'fastmusic954': 'Bad Request'}]

In [11]:
#combine 
import pandas as pd
from pathlib import Path

p = Path('feature')
folders = [x for x in p.iterdir() if x.is_dir() and x.stem.startswith("buy")]

full_df = pd.DataFrame()
for data_dir in folders:
    for parquet_file in data_dir.glob('*.parquet'):
        ig_username = parquet_file.stem
        df = pd.read_parquet(parquet_file)
        df['ig_username'] = ig_username
        full_df = pd.concat([full_df, df])

In [32]:
full_df

NameError: name 'full_df' is not defined

## Preprocess Features

In [13]:
from emot.emo_unicode import UNICODE_EMO, EMOTICONS
import re

def lower_text(text):
    try:
        text = text.lower()
    except:
        return ""
    return text
def remove_newline(string):
    string = string.replace("\n", "")
    return string
def remove_punctuation(string): 
    punctuations = '''()-[]{};:'"\<>/@#$%^&*~'''
    for x in string.lower(): 
        if x in punctuations: 
            string = string.replace(x, "")
        elif x == ",._":
            string = string.replace(x, " ")
        elif x == "?":
            string = string.replace(x, "[question mark]")
    return string

def convert_emojis(text):
    for emot in UNICODE_EMO:
        text = text.replace(emot, "_".join(UNICODE_EMO[emot].replace(",","").replace(":","").split()))
    return text

def convert_emoticons(text):
    for emot in EMOTICONS:
        text = re.sub(u'('+emot+')', "_".join(EMOTICONS[emot].replace(",","").split()), text)
    return text

def combine_metadata(row):
    return f"caption {row['caption']}, is of type {row['media_type'].lower()}, has {row['comments_count']} comments and {row['like_count']} likes"


In [14]:
full_df['full_caption'] = full_df.apply(combine_metadata, axis=1)
full_df['full_caption'] = full_df['full_caption'].apply(lower_text)
full_df['full_caption'] = full_df['full_caption'].apply(remove_newline)
full_df['full_caption'] = full_df['full_caption'].apply(convert_emojis)
full_df['full_caption'] = full_df['full_caption'].apply(convert_emoticons)
full_df_concat = full_df[['ig_username','full_caption']].groupby("ig_username").sum()

In [15]:
full_df_concat

full_caption
ig_username                                                         
2jpantoja          caption some p-bar flow from today. this was t...
_moneybo           caption wait on it real father celebrate diffe...
_ts97__            caption none, is of type video, has 0 comments...
aless_capparelli   caption firepozor nová mega súťažfire.žrebova...
alexsaenz          caption the joy of life is rarely about the de...
...                                                              ...
vngnc              caption short little breakdown of project cerb...
waynebutlercomedy  caption december 23, 2004 was the day my life ...
woolimusic         caption wooli “id collection”our biggest merch...
zairmontes         caption mi #tbt de hoy es actuando… interpreta...
zambezijuice       caption bflowmusic ‘s long-awaited album "sund...

[88 rows x 1 columns]

In [17]:
full_df_concat.to_parquet("buy_insta_accounts_inf_first100.parquet")

## These are used particularly for training data prep

In [16]:
columns = ['username', 'Openness','Conscientiousness','Extroversion','Agreeableness','Neuroticism']
feature_df =full_df_concat.merge(gc_df, left_index=True, right_on="username", how="left")
gc_df = pd.concat([
    df_buy_insta_accounts[columns],
    df_buy_biz[columns],
    df_buy_insta_accounts_add_on[columns]
])

NameError: name 'gc_df' is not defined

In [51]:
feature_df

full_caption          username  \
59    caption did you know henry ford made the 8 hou...            36mktg   
69    caption so clean smiling_face_with_heart-eyesr...    6staroileaters   
26    caption through the night crescent_moon • maet...        740project   
57    caption take your social media accounts out of...  _boostfruit.com_   
10    caption appreciate you excellencealways folded...      _cloutfinder   
...                                                 ...               ...   
71    caption none, is of type image, has 1 comments...      wipmarketing   
14    caption enjoy this video while i disappear for...         xcarmeeen   
5461  caption power back.#fanggang | xfl, is of type...         xflvipers   
67    caption summeeeeer! all the hard work of clean...       yogabody_uk   
84    caption now booking grammy award winning artis...          zgroupla   

      Openness  Conscientiousness  Extroversion  Agreeableness  Neuroticism  
59         1.0                1.0           1.0            1.0          0.0  
69         0.0                1.0           0.5            0.5          0.5  
26         1.0                0.0           0.0            0.0          0.5  
57         1.0                0.0           1.0            0.5          0.5  
10         1.0                0.0           1.0            0.5          0.5  
...        ...                ...           ...            ...          ...  
71         1.0                1.0           1.0            1.0          0.0  
14         0.0                0.0           1.0            0.0          1.0  
5461       0.0                1.0           1.0            1.0          0.0  
67         1.0                1.0           1.0            1.0          0.0  
84         1.0                0.0           1.0            0.5          0.5  

[269 rows x 7 columns]

In [50]:
feature_df['Openness'] = feature_df['Openness'].map({"High":1.0,"Low":0,"Vague":0.5})
feature_df['Conscientiousness'] = feature_df['Conscientiousness'].map({"High":1.0,"Low":0,"Vague":0.5})
feature_df['Extroversion'] = feature_df['Extroversion'].map({"High":1.0,"Low":0,"Vague":0.5})
feature_df['Agreeableness'] = feature_df['Agreeableness'].map({"High":1.0,"Low":0,"Vague":0.5})
feature_df['Neuroticism'] = feature_df['Neuroticism'].map({"High":1.0,"Low":0,"Vague":0.5})

In [94]:
feature_so_far = feature_so_far.dropna()

In [52]:
feature_df.to_parquet("feature.parquet")

In [37]:
from google.cloud import storage

client = storage.Client()
bucket = client.get_bucket('feature')
    

Forbidden: 403 GET https://storage.googleapis.com/storage/v1/b/feature?projection=noAcl&prettyPrint=false: 100822155188-compute@developer.gserviceaccount.com does not have storage.buckets.get access to the Google Cloud Storage bucket. Permission 'storage.buckets.get' denied on resource (or it may not exist).